content/precise-types.md

```diff
@@ -0,0 +1,517 @@
+16 |val v3T: Vec[3] = v3 // error
+   |                  ^^                                         
+   |                  Found:    (v3 : Vec[Int + Int])
+   |                  Required: Vec[(3 : Int)]
+~~~
+Implicit summoning of a covariant type parameter causes widening. To avoid widening here we can define the type class `TC` as invariant (and modify the related given declarations), or drop the covariance from `Vec`, or add a `Singleton` upper-bound to `S0`.
+
+As seen from all the examples thus far, the scala compiler is very eager to widen types in various cases and we need a mechanism to prevent this when the compile-time information preservation is crucial.
+
+## Proposed solution
+
+### High-level overview
+
+We propose to add a `precise` modifier that can be applied on type parameters to indicate to the compiler that term arguments annotated with those precise types are considered to be *precisely typed*. Precise typing is not only preventing widening for literals as demonstrated so far, but also prevents any widening in the expression composition, so that composable expressions like tuples can preserve their precise representation throughout their entire structure.
+
+Here is our earlier implementation of `Vec`, modified to work with all examples by applying the `precise` modifier:
+~~~ scala
+class Vec[precise +S <: Int](val size: S):
+  def ++[precise TS <: Int](that: Vec[TS]): Vec[S + TS] = 
+    Vec[S + TS]((size + that.size).asInstanceOf[S + TS])
+~~~
+The `precise` modifier helps preserving the precise size type when constructing a vector (precise typing of the size term argument), and when the vector is applied as an argument to the concatenation (not widening the covariant size type parameter).
+
+
+### Specification
+
+#### Add The `precise` Modifier
+The precise modifier is a soft keyword that is only available at a type parameter declaration position.
+The following grammar rules are modified to support this new soft keyword:
+~~~
+ClsTypeParam ::=  {Annotation} [‘precise’] [‘+’ | ‘-’] id [HkTypeParamClause] TypeParamBounds ;
+DefTypeParam ::=  {Annotation} [‘precise’] id [HkTypeParamClause] TypeParamBounds ;
+TypTypeParam ::=  {Annotation} [‘precise’] id [HkTypeParamClause] TypeBounds ;
+HkTypeParam  ::=  {Annotation} [‘precise’] [‘+’ | ‘-’] (id [HkTypeParamClause] | ‘_’) TypeBounds ;
+~~~ 
+
+#### Precise Types
+A type is considered to be precise when at least of one of following occurs:
+  * It is a type parameter declaration that is directly modified with `precise` 
+  ~~~ scala
+  //`T` is a precise type parameter declaration
+  class PreciseBox[precise T]
+  ~~~
+  * It is a type variable reference with a precise type parameter origin
+  ~~~ scala
+  //`: T` type variable is precise because it has the precise origin type parameter `T` 
+  def id[precise T](t: T): Box[T] = ??? 
+  val x = id(1) //: Box[1]
+  ~~~
+  * It is substituting a precise type parameter (may also be a wildcard substitution)
+  ~~~ scala
```

Comment:
I think you will find that this will be really hard to specify and implement correctly.

Reply:
> I think you will find that this will be really hard to specify and implement correctly.

Let's assume we can. There are far more complex things in the compiler inference system.
You can review the attempted covering of various feature interactions in the implementation:
https://github.com/lampepfl/dotty/blob/1824ad827f620b6946d56fcbeeeb5a1fe340f26c/tests/neg/precise-typecheck.scala
(there are additional test files, but this is the primary one)